### PR TITLE
gRPC Backend Plugins

### PR DESCRIPTION
**Pull request by briankassouf**

### CODE REVIEW

File: logical/plugin/grpc_backend_client.go [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+		return false, false, pb.ProtoErrToErr(reply.Err)
+	}
+
+	return reply.CheckFound, reply.Exists, nil
+}
+
+func (b *backendGRPCPluginClient) Cleanup() {
+	// Timout the connection incase we have a bad connection. We can't block on
+	// shutdown.
+	ctx, cancel := context.WithTimeout(b.doneCtx, time.Second)
```
Comment (jefferai) [inline logical/plugin/grpc_backend_client.go]: One second feels short. I could imagine that a plugin might need to close a number of connections or finalize resources.

File: logical/plugin/grpc_backend_client.go [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+	"google.golang.org/grpc/connectivity"
+
+	"github.com/hashicorp/go-plugin"
+	"github.com/hashicorp/vault/helper/pluginutil"
+	"github.com/hashicorp/vault/logical"
+	"github.com/hashicorp/vault/logical/plugin/pb"
+	log "github.com/mgutz/logxi/v1"
+)
+
+var ErrPluginShutdown = errors.New("plugin is shutdown")
```
Comment (calvn) [inline logical/plugin/grpc_backend_client.go]: s/shutdown/shut down

File: logical/plugin/grpc_backend_client.go (hunk 1 above)
Comment (briankassouf) [reply logical/plugin/grpc_backend_client.go]: this should actually be removed and replaced with doneCtx. Will fix

File: logical/plugin/backend.go [hunk 3]
```diff
@@ -22,3 +29,23 @@ func (b *BackendPlugin) Server(broker *plugin.MuxBroker) (interface{}, error) {
...
 	return &backendPluginClient{client: c, broker: broker, metadataMode: b.metadataMode}, nil
 }
+
+func (b BackendPlugin) GRPCServer(broker *plugin.GRPCBroker, s *grpc.Server) error {
+	pb.RegisterBackendServer(s, &backendGRPCPluginServer{
+		broker:  broker,
+		factory: b.Factory,
+		// We pass the logger down into the backend so go-plugin will forward
+		// logs for us.
+		logger: logbridge.NewLogger(b.Logger).LogxiLogger(),
```
Comment (calvn) [inline logical/plugin/backend.go]: Where do we set `b.Logger`? I couldn't find any reference on where it's being set.

Comment (briankassouf) [reply logical/plugin/backend.go]: https://github.com/hashicorp/vault/pull/3808/files/bc31a4623c998e6baa66544534ccb14a90bd8ae9#diff-046797d040beeb3fd5d9ff2afedcefc6R42

**Pull request merged**